From 383c0ee3af0270cb03ccfd8443b3b04f7b52ab90 Mon Sep 17 00:00:00 2001
From: Yucheng Lu <kanolyc@gmail.com>
Date: Thu, 30 Apr 2026 07:29:28 +0000
Subject: [PATCH 65/65] net/sched: sch_netem: fix out-of-bounds access in
 packet corruption

In netem_enqueue(), the packet corruption logic uses
prandom_u32() % skb_headlen(skb) to select an index for
modifying skb->data. When an AF_PACKET TX_RING sends fully non-linear
packets over an IPIP tunnel, skb_headlen(skb) evaluates to 0.

Passing 0 as the modulus operand results in undefined behavior (integer
division by zero), with the resulting unconstrained random value used as
an offset into skb->data causing an out-of-bounds memory access.

Fix this by verifying skb_headlen(skb) is non-zero before attempting
to corrupt the linear data area. Fully non-linear packets will silently
bypass the corruption logic.

Fixes: c865e5d99e25 ("[PKT_SCHED] netem: packet corruption option")
Reported-by: Yifan Wu <yifanwucs@gmail.com>
Reported-by: Juefei Pu <tomapufckgml@gmail.com>
Signed-off-by: Yuan Tan <tanyuan98@outlook.com>
Signed-off-by: Xin Liu <bird@lzu.edu.cn>
Signed-off-by: Yuhang Zheng <z1652074432@gmail.com>
Signed-off-by: Yucheng Lu <kanolyc@gmail.com>
Reviewed-by: Stephen Hemminger <stephen@networkplumber.org>
Link: https://patch.msgid.link/45435c0935df877853a81e6d06205ac738ec65fa.1774941614.git.kanolyc@gmail.com
Signed-off-by: Jakub Kicinski <kuba@kernel.org>
(cherry picked from commit d64cb81dcbd54927515a7f65e5e24affdc73c14b)
CVE: CVE-2026-31675
(cherry picked from commit 82964f5cf0ab0b0f5ce7a1d2ed7868b52fa60f54)
---
 net/sched/sch_netem.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/net/sched/sch_netem.c b/net/sched/sch_netem.c
index 951156d7e5485..a16535b584b9c 100644
--- a/net/sched/sch_netem.c
+++ b/net/sched/sch_netem.c
@@ -508,8 +508,9 @@ static int netem_enqueue(struct sk_buff *skb, struct Qdisc *sch,
 			goto finish_segs;
 		}
 
-		skb->data[prandom_u32() % skb_headlen(skb)] ^=
-			1<<(prandom_u32() % 8);
+		if (skb_headlen(skb))
+			skb->data[prandom_u32() % skb_headlen(skb)] ^=
+				1<<(prandom_u32() % 8);
 	}
 
 	if (unlikely(q->t_len >= sch->limit)) {
-- 
2.43.0

